aerial_robot_nerve/spinal/mcu_project/Jsk_Lib/flight_control/attitude/attitude_control.cpp

```diff
@@ -407,11 +442,203 @@ void AttitudeController::reset(void)
+    mrac_target_cog_torque_[axis] = -mrac_phi_gamma[axis]-mracControlVal[axis];
+    mrac_ref_model_msg_.data[axis] = rot_ref_[axis];
+  }
+
+  // limit yaw torque
+  mrac_target_cog_torque_[Z] = limit(mrac_target_cog_torque_[Z], YAW_TORQUE_LIMIT);
+
+  // integral mrac parameter update
+  rot_ref_ = rot_ref_ + rot_ref_d_ * DELTA_T;
+  rot_ref_d_ = rot_ref_d_ + (k1m_*(rot_des - rot_ref_) + k2m_*(rot_des_d - rot_ref_d_)) * DELTA_T;
+  ap::Vector3f edotPlusK1e = e_ref_dot + K1_ * e_ref;
+  float gamma_dot[8];
+  gamma_dot[0] = kappa_[0] * ( edotPlusK1e[0]*p[0] );
+  gamma_dot[1] = kappa_[1] * ( edotPlusK1e[1]*p[1] );
+  gamma_dot[2] = kappa_[2] * ( edotPlusK1e[2]*p[2] );
+  gamma_dot[3] = kappa_[3] * ( edotPlusK1e[0]*p[1] + edotPlusK1e[1]*p[0] );
+  gamma_dot[4] = kappa_[4] * ( edotPlusK1e[1]*p[2] + edotPlusK1e[2]*p[0] );
+  gamma_dot[5] = kappa_[5] * ( edotPlusK1e[0]*p[2] + edotPlusK1e[2]*p[1] );
+  gamma_dot[6] = kappa_[6] * ( edotPlusK1e[0] );
+  gamma_dot[7] = kappa_[7] * ( edotPlusK1e[1] );
+  for (int i=0; i<8 ; i++) {
+    gamma_[i] = gamma_[i] + (gamma_dot[i] -sigma_[i]*gamma_[i])* DELTA_T;
+    // make sure diagonal of inertia is positive
+    if (i<3) {
+      if (gamma_[i] < 0) {
+        gamma_[i] = 0;
+      }
+    }
+    mrac_gamma_msg_.data[i] = gamma_[i];
+  }
+
+  if (mrac_param_pub_control_ > mrac_log_rate_) {
+#ifdef SIMULATION
+    mrac_gamma_pub_.publish(mrac_gamma_msg_);
+    mrac_ref_model_pub_.publish(mrac_ref_model_msg_);
+#else
+    mrac_gamma_pub_.publish(&mrac_gamma_msg_);
+    mrac_ref_model_pub_.publish(&mrac_ref_model_msg_);
+#endif
+    mrac_param_pub_control_ = 0;
+  } else {
+    mrac_param_pub_control_++;
+  }
+}
+
+#ifdef SIMULATION
+bool AttitudeController::setMRACTriggerCallback(std_srvs::SetBool::Request& req, std_srvs::SetBool::Response& res)
+#else
+void AttitudeController::setMRACTriggerCallback(const std_srvs::SetBool::Request& req, std_srvs::SetBool::Response& res)
+#endif
+{
+  is_use_mrac_ = req.data;
+  if (is_use_mrac_) {
+    mrac_reset_flag_ = true;
+  }
+  res.success = true;
+#ifdef SIMULATION
+  return true;
+#endif
+}
+
+#ifdef SIMULATION
+bool AttitudeController::setMRACParamsCallback(spinal::SetMRACParams::Request& req, spinal::SetMRACParams::Response& res)
+#else
+void AttitudeController::setMRACParamsCallback(const spinal::SetMRACParams::Request& req, spinal::SetMRACParams::Response& res)
+#endif
+{
+  mrac_log_rate_ = req.log_rate;
+  for (int axis=0; axis<3 ; axis++) {
+    mrac_ratio_[axis] = req.mrac_ratio[axis];
+    k1m_[axis][axis] = req.k1m[axis];
+    k2m_[axis][axis] = req.k2m[axis];
+    K1_[axis][axis] = req.K1[axis];
+    K2_[axis][axis] = req.K2[axis];
+  }
+  for (int i=0; i<8 ; i++) {
+    kappa_[i] = req.kappa[i];
+    sigma_[i] = req.sigma[i];
+  }
+  attitude_p_ = req.attitude_p;
+  res.success = true;
+#ifdef SIMULATION
+  return true;
+#endif
+}
+
+void AttitudeController::yawFromPCallback(const spinal::YawFromPC& msg)
+{
+  pc_psi_ = msg.pc_yaw;
```

Comment:
Ok, so you need the desired yaw angle and the current yaw angle.
The former one can be merged to `FourAxisCommand.msg`.
The later one should be sent independently.
Please at least change the message name `yawFromPC` to only `yaw` or `estimatedYaw`, not `FromPC`.